### PR TITLE
fix(react-hooks): mark attribute filtering for `useActive` and `useAttrs`

### PR DESCRIPTION
**Pull request by whawker**
### Description

Add support for attribute filtering for `useActive` and `useAttrs` hooks when used with marks.

This provides consistent behaviour for the hook, aligning with functionality provided for node types.

```tsx
const active = useActive();

// Previously this ignored passed attributes and only checked the mark's type
//
// Now this will only return true if mark type is active AND its color attribute is red
const isActive = active.textColor({ color: 'red' });
```

This is added as a **fix** rather than a **feature**, as [types indicate this is the expected behaviour](https://github.com/remirror/remirror/blob/main/packages/remirror__core/src/extension/extension-types.ts#L268).

### Checklist

- [x] I have read the [**contributing**](https://github.com/remirror/remirror/blob/HEAD/docs/contributing.md) document.
- [x] My code follows the code style of this project and `pnpm fix` completed successfully.
- [x] I have updated the documentation where necessary.
- [x] New code is unit tested and all current tests pass when running `pnpm test`.

### Demo

[Text color Storybook](https://pr1510-remirror-ocavue.vercel.app/?path=/story/extensions-textcolor--basic)

### CODE REVIEW

**Review by ankon — APPROVED** [review]
LGTM, added some small notes about stuff I noticed when reading through it.

File: packages/remirror__core-utils/src/core-utils.ts [hunk 1]
```diff
@@ -235,6 +237,52 @@ export function isResolvedPos(value: unknown): value is PMResolvedPos<EditorSche
...
+ * @param props - see [[`RangeHasMarkProps`]] for options
+ */
+export function rangeHasMark(props: RangeHasMarkProps): boolean {
+  const { trState, from, to, type, attrs } = props;
+  const { doc } = trState;
+  const markType = getMarkType(type, doc.type.schema);
+
+  const attrEntries = Object.entries(attrs ?? {});
+
+  if (!attrs || attrEntries.length === 0) {
```
[the remark this answers is not in the document]
Comment (whawker) [reply packages/remirror__core-utils/src/core-utils.ts:258]: This is to make typescript happy for the `containsAttributes` call on line 276

File: packages/remirror__core-utils/src/prosemirror-utils.ts [hunk 2]
```diff
@@ -749,3 +751,24 @@ export function mergeProsemirrorKeyBindings<Schema extends EditorSchema = Editor
...
+): boolean {
+  const currentAttrs = nodeOrMark.attrs ?? {};
+
+  for (const [name, value] of Object.entries(attrs)) {
+    if (currentAttrs[name] !== value) {
+      return false;
+    }
+  }
+
+  return true;
```
[the remark this answers is not in the document]
Comment (whawker) [reply packages/remirror__core-utils/src/prosemirror-utils.ts:773]: My thought was with `for...of` we can bail out as soon as we have a mismatch, but I suppose `every` would shortcut in the same way too.

Make sense though, will add.

File: packages/remirror__core-utils/src/core-utils.ts [hunk 3]
```diff
@@ -254,7 +306,7 @@ interface IsMarkActiveProps extends MarkTypeProps, Partial<FromToProps>, TrState
  * @param props - see [[`IsMarkActiveProps`]] for options
  */
 export function isMarkActive(props: IsMarkActiveProps): boolean {
-  const { trState, type, from, to } = props;
+  const { trState, type, attrs, from, to } = props;
```
[the remark this answers is not in the document]
Comment (whawker) [reply packages/remirror__core-utils/src/core-utils.ts:309]: This was to done to allow a quick bail out to the default function [here](https://github.com/remirror/remirror/blob/061b9431b7b15057b75b52b85babeb50c957d069/packages/remirror__core-utils/src/core-utils.ts#L258-L260).

But I suppose `attrsEntries.length === 0` would also catch that case, if we supplied the default here.

File: packages/remirror__core/src/builtins/helpers-extension.ts [hunk 4]
```diff
@@ -88,12 +89,22 @@ export class HelpersExtension extends PlainExtension {
...
+
+          if (!markRange || !attrs) {
+            return markRange?.mark.attrs;
+          }
+
+          if (containsAttributes(markRange.mark, attrs)) {
+            return markRange.mark.attrs;
+          }
+
+          return;
```
[the remark this answers is not in the document]
Comment (whawker) [reply packages/remirror__core/src/builtins/helpers-extension.ts:107]: This was a compromise to make both typescript and eslint happy.

Typescript was complaining that "not all code paths return a value"
Eslint was complaining about "no useless return"

File: packages/remirror__core-utils/src/core-utils.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (ankon) [reply packages/remirror__core-utils/src/core-utils.ts:309]: Fair point, if the `attrs is sometimes undefined` is used, then that makes sense for me.

File: packages/remirror__core-utils/src/core-utils.ts [hunk 5]
```diff
@@ -235,6 +237,52 @@ export function isResolvedPos(value: unknown): value is PMResolvedPos<EditorSche
...
+
+  const attrEntries = Object.entries(attrs ?? {});
+
+  if (!attrs || attrEntries.length === 0) {
+    return doc.rangeHasMark(from, to, markType);
+  }
+
+  let found = false;
+
+  if (to > from) {
```
[the remark this answers is not in the document]
Comment (whawker) [reply packages/remirror__core-utils/src/core-utils.ts:262]: Added

https://github.com/remirror/remirror/blob/2ce8f4a706a890324d872abc26d31c502692e473/packages/remirror__core-utils/__tests__/core-utils.spec.ts#L107-L113

https://github.com/remirror/remirror/blob/2ce8f4a706a890324d872abc26d31c502692e473/packages/remirror__core-utils/__tests__/core-utils.spec.ts#L195-L201

**Pull request merged**